### PR TITLE
Framework: Replace element-closest with registered vendor script

### LINKED ISSUE #7159: Framework: Use standard polyfill pattern for Element#closest polyfill
Related: #7033 (specifically overview at https://github.com/WordPress/gutenberg/pull/7033#pullrequestreview-124835640 )

We should consolidate our approaches to polyfilling missing behavior, mostly impacting Internet Explorer support. Our use of the `element-closest` module has proven difficult to maintain. The approach used in #7033 for polyfilling `Node#contains` is looking to be a better candidate for its maintenance and performance implications.

As part of this, we must determine a way to ensure these polyfills are loaded by all dependent scripts. Some ideas:

>In that sense, and given that these are true progressive enhancements which do nothing if a browser is detected to support a given feature, I'd say: Load it everywhere; every WordPress admin screen; for any features we ever use.

https://github.com/WordPress/gutenberg/pull/7033#pullrequestreview-124835640 

>we should probably not need to load it if:
>
>- There are zero scripts on a screen otherwise
>- We have some way of knowing / inferring that modern authoring practices are employed
>   - Only those which are transpiled (from a build directory)?
>   - Some other metric akin to @babel/preset-env@next's useBuiltIns: 'usage'?
>   - Register scripts through a new abstracted registration function, e.g. wp_register_script_with_polyfill?

https://github.com/WordPress/gutenberg/pull/7033#issuecomment-393572986

>There's a couple things I'd like to see, but not necessarily as part of these changes:
>
>- Avoid polyfills being considered as dependencies of specific scripts, unless we're certain we can maintain those dependencies.
>   - There's nothing about wp-dom that depends on the Node#contains polyfill. Wherever we're using it should be the dependency
>   - But then again, this highlights the bigger problem that this is a source of maintenance headaches and is likely not worth maintaining, vs. load always
>- Better names for the script enqueues, which avoids potential plugin incompatibilities. Probably something with a WP-specific prefix, e.g. wp-polyfill-node-contains

https://github.com/WordPress/gutenberg/pull/7033#pullrequestreview-125293992

### PR DESCRIPTION
**Pull request by aduth**
Closes #7159

This pull request seeks to eliminate `element-closest` dependency and imports from individual packages, instead replacing with a standard polyfill approach via server-side script enqueue.

Potential future improvements include:

- Improved association between script handle as being a polyfill (avoid `wp_add_inline_script` per dependent)
- Avoid polyfill dependencies needing to be kept in sync with current state of codebase using / not using (maybe always load)

**Testing instructions:**

Verify there are no regressions in `Element#closest` behaviors, notably in IE11 where it is not supported by default. For example, verify that inserting a quote block and ArrowDown'ing through its citation and into the addition of a default block (verifies usage of `Element#closest` in WritingFlow's use of `isInSameBlock`).

### CODE REVIEW

**Comment by youknowriad** [comment]
I wonder if we should just do it for `wp-polyfill-ecmascript` (adding the dependency). This would avoid us to keep track of the exact usage of this feature per script which is very hard to accomplish anyway.

**Comment by aduth** [comment]
@youknowriad Yes, I wasn't too sure how we might want to accomplish the "load always" behavior for `element-closest`. There's a bit of additional context in #7159 and #7033. With your suggestion, we'd expand what's implied by adding a dependency on `wp-polyfill-ecmascript` from meaning "latest JavaScript features" to also including "modern browser behaviors". These often collide anyways in our usage, and combining them would make for a simpler developer experience, but they do hold separate meanings which could be worth distinguishing.

Maybe a topic for tomorrow's chat? 🙂

**Comment by aduth** [comment]
Put another way, ECMAScript and WHATWG/W3 standards are separate, though often used interchangeably as we use modern features and for polyfilling. Should we or shouldn't we abandon attempts to keep them distinct?

**Comment by youknowriad** [comment]
`core-js` which is basically what `babel-polyfill` uses is not only about `ECMAScript` but it also includes some `WHATWG/W3` standards. I'm not really certain if `babel-polyfill` picks only the `ECMAScript` features thought or just enqueues the whole thing.

**Comment by aduth** [comment]
>core-js which is basically what babel-polyfill uses is not only about ECMAScript but it also includes some WHATWG/W3 standards.

Is this true? Could you name some such features?

Genuinely curious because I'd commented as such that my assumption is that it's only for ECMAScript:

https://github.com/WordPress/gutenberg/pull/9171#issuecomment-419536905

**Comment by aduth** [comment]
Oh I see now from their documentation they have a separate section for web standards:

https://github.com/zloirock/core-js#features

Then it becomes a question of: Are we aligned to `core-js`, or to `babel-polyfill`? And does `babel-polyfill` make the same commitments?

**Comment by aduth** [comment]
Rebased to account for the now-merged #9794, which simplifies the implementation here a fair bit.

I also discovered that newer versions of JSDOM include support for `Element#closest` (https://github.com/jsdom/jsdom/issues/1555) such that it wouldn't be necessary to explicitly provide this in `jest-preset-default`. See f6916bd.

In updating our pinned version, however, I fought with the license checker flagging `abab`. I think we've seen this on a few occasions before. Ultimately I decided to add an `--ignore` option to the license checker, to be used for these sorts of cases where we can manually vet the license for a package as being compatible. The `abab` package has a compatible license (W3C BSD 3-Clause), but because [it is specified in its `package.json` in a non-parseable format](
https://github.com/jsdom/abab/blob/4327de3aae348710094d9f3c1f0c1477d9feb865/package.json#L26), it is marked as being invalid. See 674e70f.

**Comment by aduth** [comment]
cc @pento re: changes to `wp-scripts check-licenses`

**Comment by pento** [comment]
Licenses are the worst. I mean, I get why they don't put an identifier in the `license` field, because there isn't a SPDX identifier for the W3C variant of the BSD 3-Clause license. But they could just make it up.

I like the `--ignore` option, but I hope we never have to use it. 65bf762 adds `BSD-3-Clause-W3C` as a GPL2 compatible license, and adds a check for if the `license` field is telling us to look in a license file, instead.

File: packages/scripts/scripts/check-licenses.js [hunk 1]
```diff
@@ -193,6 +225,11 @@ modules.forEach( ( path ) => {
 		);
 	let licenseType = typeof license === 'object' ? license.type : license;
 
+	// Check if the license we've detected is telling us to look in the license file, instead.
+	if ( licenseType && licenseFiles.find( ( licenseFile ) => licenseType.indexOf( licenseFile ) >= 0 ) ) {
+		licenseType = undefined;
```
Comment (aduth) [inline packages/scripts/scripts/check-licenses.js]: What if the `license` property is pointing a file describing a truly incompatible license? Wouldn't this wrongly allow the license to be considered valid? I expect this to be even more common for non-standard licenses.

Comment (aduth) [reply packages/scripts/scripts/check-licenses.js]: Oh! I missed the code immediately following this, where we try to infer the license type from the presence of a file.

**Comment by aduth** [comment]
Rebased to resolve conflicts.

At this point just need a 👍

File: packages/scripts/scripts/check-licenses.js (hunk 1 above)
Comment (pento) [reply packages/scripts/scripts/check-licenses.js]: 🙂

**Review by brandonpayton — APPROVED** [review]
I tested this in IE11, and everything appears to work as expected. I also checked this in Chrome and observed that the native version of `HTMLElement.prototype.closest` was intact.

👍  LGTM

File: packages/scripts/scripts/check-licenses.js [hunk 2]
```diff
@@ -193,6 +225,11 @@ modules.forEach( ( path ) => {
 		);
 	let licenseType = typeof license === 'object' ? license.type : license;
 
+	// Check if the license we've detected is telling us to look in the license file, instead.
+	if ( licenseType && licenseFiles.find( ( licenseFile ) => licenseType.indexOf( licenseFile ) >= 0 ) ) {
```
Comment (brandonpayton) [inline packages/scripts/scripts/check-licenses.js]: nit: This is a good place to use `String#includes` which reads a bit better than an index comparison.

**Review by mcsf — APPROVED** [review]
Good change. :)

**Pull request merged**